### PR TITLE
Clean up AIP-136: Custom methods.

### PR DESCRIPTION
**Pull request by rofrankel**
This is in preparation for the barn raising event, and makes the AEP more consistent with more recently-adopted AEPs.  Common guidance is also factored out of the IDL-specific tabs.

I added one new bullet point of guidance while I was at it.  I doubt this will be controversial, but calling it out for completeness.

I actually think most of the changes from the Google AIP are pretty good in this case, so I have generally not modified them.

### CODE REVIEW

File: aep/general/0136/aep.md.j2
Comment (toumorokoshi) [inline aep/general/0136/aep.md.j2:44]: why not must?

File: aep/general/0136/aep.md.j2
Comment (toumorokoshi) [inline aep/general/0136/aep.md.j2:46]: same as above.

File: aep/general/0136/aep.md.j2
Comment (toumorokoshi) [inline aep/general/0136/aep.md.j2:83]: why no resource-based custom methods? https://google.aip.dev/136#resource-based-custom-methods

File: aep/general/0136/aep.md.j2
Comment (toumorokoshi) [inline aep/general/0136/aep.md.j2:108]: This is guidance on the "management plane" resources really - although maybe we need to work on some nomenclature here.

I'd suggest either:

1. calling this "management plane resources" match the other AIPs (e.g. the create you're adding)
2. remove this guidance for now, and add in a follow-up PR once we get things ironed out.

thoughts?

**Review by toumorokoshi — CHANGES_REQUESTED** [review]
There's a lot missing from this AEP that was added in the AIP: I think I pointed out the important ones in the comments, but we should add those in or decide what to do with those before merging.

File: aep/general/0136/aep.md.j2
Comment (mkistler) [inline aep/general/0136/aep.md.j2:62]: Let's go back to the Google guidance here.

> - The HTTP method must be GET or POST:
>   - GET must be used for methods retrieving data or resource state.
>   - POST must be used if the method has side effects or mutates resources or data.
> - The HTTP URI must use a : character followed by the custom verb (:archive in the above example), and the verb in the URI must match the verb in the name of the RPC.
>   - If word separation is required, camelCase must be used.

**Review by mkistler — APPROVED** [review]
This mostly looks fine, but there is one section of the Google guidance that was dropped that I think should be restored.

I'll approve now to avoid a re-review before merging, unless there is disagreement on restoring the Google guidance as I requested.

File: aep/general/0136/aep.md.j2
Comment (rofrankel) [reply aep/general/0136/aep.md.j2:19]: Discussed live:

* Rich is concerned that making this a **must** without any kind of informal proof that there won't be a valid use case for e.g. DELETE risks someone running up against the AEPs not supporting a valid use case.  (E.g. delete this node, reparent its children, and cache the new parent for <duration> for nodes newly parented to the deleted node.)
* Yusuke thinks this guidance lacks context; people may just implement patch semantics with a POST request.
* Conclusion:  refer to the IETF document on HTTP methods and say to follow those.  RFC 7231 (and whichever one has PATCH).  Demote this guidance to a "in practice, most of the time" style note.

File: aep/general/0136/aep.md.j2
Comment (rofrankel) [reply aep/general/0136/aep.md.j2:21]: * Rich:  real example [here](https://create.roblox.com/docs/en-us/cloud/reference/Instance) (`:listInstanceChildren`)
* Yusuke:  what if we split the List AEP in two and then have the first refer to the second?
  * GET for collections
  * List methods
* We'll take this out for now; Rich will file an issue to revisit this later.

File: aep/general/0136/aep.md.j2
Comment (rofrankel) [reply aep/general/0136/aep.md.j2:44]: Another issue:  "matching" is underspecified.

Can we make this guidance more precise, e.g. that it must be the RPC name + "Request"?

Or do we not care at all, because it's a custom method?

In general, it's nice for **should**s to have a counterexample.

Conclusion:  change this to **must** and clarify "matching"; we can always change it back to **should** if we find a good counterexample.

File: aep/general/0136/aep.md.j2
Comment (rofrankel) [reply aep/general/0136/aep.md.j2:108]: Discussed live:  Yusuke suggests just leaving it after all.

File: aep/general/0136/aep.md.j2
Comment (rofrankel) [reply aep/general/0136/aep.md.j2:12]: Rephrased, though I'm concerned about the possibility of debate between "if possibe" in the previous sentence and "contort" in this one.  Maybe we just drop this sentence?

File: aep/general/0136/aep.md.j2
Comment (rofrankel) [reply aep/general/0136/aep.md.j2:46]: In this case, they might return the resource itself instead - see the sub-bullet.

File: aep/general/0136/aep.md.j2
Comment (rofrankel) [reply aep/general/0136/aep.md.j2:19]: Done.

File: aep/general/0136/aep.md.j2
Comment (rofrankel) [reply aep/general/0136/aep.md.j2:62]: For the first bullet point, see discussion [here](https://github.com/aep-dev/aep.dev/pull/103/files#r1518530318).

Second bullet point reverted to Google style.

File: aep/general/0136/aep.md.j2
Comment (rofrankel) [reply aep/general/0136/aep.md.j2:44]: Done.

File: aep/general/0136/aep.md.j2
Comment (rofrankel) [reply aep/general/0136/aep.md.j2:21]: Removed and filed https://github.com/aep-dev/aep.dev/issues/112.

File: aep/general/0136/aep.md.j2
Comment (rofrankel) [reply aep/general/0136/aep.md.j2:97]: Done.

File: aep/general/0136/aep.md.j2
Comment (toumorokoshi) [inline aep/general/0136/aep.md.j2:13]: FWIW "sort of work" is ambiguous - might be a place to fix later.

File: aep/general/0136/aep.md.j2
Comment (toumorokoshi) [reply aep/general/0136/aep.md.j2]: filed an issue to look at cleaning it up: https://github.com/aep-dev/aep.dev/issues/116. 

Probably best not to block this PR on this wording, we can make a surgical change on it later.

File: aep/general/0136/aep.md.j2
Comment (toumorokoshi) [reply aep/general/0136/aep.md.j2]: awesome!

File: aep/general/0136/aep.md.j2
Comment (toumorokoshi) [inline aep/general/0136/aep.md.j2:74]: write an OAS example for resource-based custom methods.

**Review by toumorokoshi — APPROVED** [review]
lgtm, thanks!

File: aep/general/0136/aep.md.j2
Comment (mkistler) [reply aep/general/0136/aep.md.j2:62]: @rofrankel the "here" link just brings me back to the files tab. Is there something more specific? Maybe copy/paste the key points here?

Comment (rofrankel) [reply aep/general/0136/aep.md.j2:62]: Oh, that's annoying, sorry.  Here's a second attempt at producing the right link: https://github.com/aep-dev/aep.dev/pull/103#discussion_r1518530318 (just worked for me when I tried it in a new tab).

But also, I'll quote it here:

> The AIP guidance says it MUST be POST or GET: I think this is a pretty valuable change that was made after the original import: https://google.aip.dev/136#guidance. Can we add it?

> Discussed live:

> * Rich is concerned that making this a must without any kind of informal proof that there won't be a valid use case for e.g. DELETE risks someone running up against the AEPs not supporting a valid use case. (E.g. delete this node, reparent its children, and temporarily cache the new parent for for nodes newly parented to the deleted node.)
> * Yusuke thinks this guidance lacks context; people may just implement patch semantics with a POST request.
> * Conclusion: refer to the IETF document on HTTP methods and say to follow those. RFC 7231 (and whichever one has PATCH). Demote this guidance to a "in practice, most of the time" style note.

Comment (rofrankel) [reply aep/general/0136/aep.md.j2:62]: Created https://github.com/aep-dev/aep.dev/issues/128 to revisit this - merging in advance of the barn raising.

**Pull request merged**
